app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,174 @@
+   * Retrieves the recoveryKeyId from uid and recoveryKey. The recoveryKeyId
+   * is a HKDF using recoveryKey as the key material and uid as the salt.
+   *
+   * @param {String} uid Uid of user
+   * @param {String} recoveryKey Recovery key
+   * @returns {Promise} A promise that will be fulfilled with the `recoveryKeyId`
+   */
+  getRecoveryKeyId: function (uid, recoveryKey) {
+    return importFxaCryptoDeriver().then(({jose}) => {
+      required(uid, 'uid');
+      required(recoveryKey, 'recoveryKey');
+
+      return Base32.decode(recoveryKey)
+        .then((keyMaterial) => {
+          keyMaterial = Buffer.from(keyMaterial, 'utf8');
+          const salt = Buffer.from(uid, 'hex');
+          const info = Buffer.from('fxa recovery fingerprint', 'utf8');
+          const options = {
+            info,
+            length: 16,
+            salt
+          };
+          return jose.JWA.derive('HKDF-SHA-256', keyMaterial, options)
+            .then((recoveryKeyIdRaw) => recoveryKeyIdRaw.toString('hex'));
+        });
+    });
+  },
+
+  /**
+   * Retrieves the recovery JWK from the uid and recovery key. The
+   * recoveryJwk can be used to encrypt and decrypt data.
+   *
+   * @param {String} uid Uid of user
+   * @param {String} recoveryKey Recovery key
+   * @returns {Promise} A promise that will be fulfilled with JWK
+   */
+  getRecoveryJwk: function (uid, recoveryKey) {
+    return importFxaCryptoDeriver().then(({jose}) => {
+      required(uid, 'uid');
+      required(recoveryKey, 'recoveryKey');
+
+      return Base32.decode(recoveryKey)
+        .then((keyMaterial) => {
+          keyMaterial = Buffer.from(keyMaterial, 'utf8');
+          const salt = Buffer.from(uid, 'hex');
+          const info = Buffer.from('fxa recovery encrypt key', 'utf8');
+          const options = {
+            info,
+            length: 32,
+            salt
+          };
+          return jose.JWA.derive('HKDF-SHA-256', keyMaterial, options)
+            .then((recoveryEncryptionRaw) => {
+              const recoveryEncryption = recoveryEncryptionRaw.toString('hex');
+              const keyOptions = {
+                alg: ENCRYPTION_ALGORITHM,
+                k: jose.util.base64url.encode(recoveryEncryption, 'hex'),
+                kty: 'oct'
+              };
+              return jose.JWK.asKey(keyOptions);
+            });
+        });
+    });
+  },
+
+  /**
+   * Creates an encrypted bundle for recovery data, which will
+   * contain the user's kB .
+   *
+   * @param {String} recoveryKeyId Recovery key id used in encryption
+   * @param {String} recoveryJwk Recovery JWK used to encrypt data
+   * @param {Object} recoveryData Data to encrypt
+   * @param {Object} [options={}] Options
+   *   @param {String} [options.iv] - Input vector used to create bundle
+   * @returns {Promise} A promise that will be fulfilled with the encrypted recoveryData
+   */
+  bundleRecoveryData: function (recoveryKeyId, recoveryJwk, recoveryData, options = {}) {
```

Comment:
If `recoveryJwk` had a `kid: recoveryKeyId`, the extra parameter would not be necessary.